File: backend/lib/legend/core/tunnel/mux.ex
defmodule Legend.Core.Tunnel.Mux do
  @moduledoc """
  Stream-multiplexing frame codec shared with the in-sprite bridge
  (`bridge/src/mux.rs` — keep both in lockstep).

  Wire: big-endian `type:u8 stream_id:u32 length:u32 payload:length`.
  Types: 1 OPEN, 2 DATA, 3 CLOSE, 4 WINDOW (payload = credit:u32).
  """

  @initial_window 262_144
  def initial_window, do: @initial_window

  @max_frame_payload 1_048_576
  @doc "Maximum accepted frame payload (1 MiB). Keep in lockstep with bridge/src/mux.rs."
  def max_frame_payload, do: @max_frame_payload

  defmodule Frame do
    @enforce_keys [:type, :stream_id]
    defstruct [:type, :stream_id, payload: ""]

    @type t :: %__MODULE__{
            type: :open | :data | :close | :window,
            stream_id: non_neg_integer(),
            payload: binary()
          }
  end

  @tag %{open: 1, data: 2, close: 3, window: 4}
  @type_of %{1 => :open, 2 => :data, 3 => :close, 4 => :window}

  @spec encode(Frame.t()) :: binary()
  def encode(%Frame{type: type, stream_id: id, payload: p}) do
    <<Map.fetch!(@tag, type), id::32, byte_size(p)::32, p::binary>>
  end

  @doc "Consume as many whole frames as `buffer` holds. {:error, :frame_too_large} aborts."
  @spec decode(binary()) :: {:ok, [Frame.t()], binary()} | {:error, :frame_too_large}
  def decode(buffer), do: decode(buffer, [])

  defp decode(<<_tag, _id::32, len::32, _rest::binary>>, _acc) when len > @max_frame_payload do
    {:error, :frame_too_large}
  end

  defp decode(<<tag, id::32, len::32, payload::binary-size(len), rest::binary>>, acc) do
    decode(rest, [%Frame{type: Map.fetch!(@type_of, tag), stream_id: id, payload: payload} | acc])
  end

  defp decode(leftover, acc), do: {:ok, Enum.reverse(acc), leftover}

  @spec window(non_neg_integer(), non_neg_integer()) :: Frame.t()
  def window(stream_id, credit),
    do: %Frame{type: :window, stream_id: stream_id, payload: <<credit::32>>}

  @spec parse_window(Frame.t()) :: {:window, non_neg_integer(), non_neg_integer()}
  def parse_window(%Frame{type: :window, stream_id: id, payload: <<credit::32>>}),
    do: {:window, id, credit}
end

File: bridge/src/mux.rs
/// Mux frame codec — must byte-match `backend/lib/legend/core/tunnel/mux.ex`.
///
/// Wire format (all big-endian, no alignment):
///   type:u8  stream_id:u32  length:u32  payload:[length bytes]
///
/// Frame types:
///   OPEN   = 1  (no payload)
///   DATA   = 2  (raw bytes)
///   CLOSE  = 3  (no payload)
///   WINDOW = 4  (payload = credit:u32 big-endian)
use std::io;
use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};

pub const OPEN: u8 = 1;
pub const DATA: u8 = 2;
pub const CLOSE: u8 = 3;
pub const WINDOW: u8 = 4;

#[allow(dead_code)]
pub const INITIAL_WINDOW: u32 = 262_144;

/// Maximum accepted frame payload (1 MiB). Keep in lockstep with mux.ex `@max_frame_payload`.
pub const MAX_FRAME_PAYLOAD: usize = 1_048_576;

/// Header size: 1 (type) + 4 (stream_id) + 4 (length) = 9 bytes.
const HEADER_LEN: usize = 9;

#[derive(Debug)]
pub struct Frame {
    pub typ: u8,
    pub stream_id: u32,
    pub payload: Vec<u8>,
}

/// Read one frame from `r`.  Returns `None` on clean EOF before the header.
pub async fn read_frame<R: AsyncRead + Unpin>(r: &mut R) -> io::Result<Option<Frame>> {
    let mut hdr = [0u8; HEADER_LEN];
    match r.read_exact(&mut hdr).await {
        Ok(_) => {}
        Err(e) if e.kind() == io::ErrorKind::UnexpectedEof => return Ok(None),
        Err(e) => return Err(e),
    }

    let typ = hdr[0];
    let stream_id = u32::from_be_bytes([hdr[1], hdr[2], hdr[3], hdr[4]]);
    let length = u32::from_be_bytes([hdr[5], hdr[6], hdr[7], hdr[8]]) as usize;

    if length > MAX_FRAME_PAYLOAD {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "frame payload exceeds MAX_FRAME_PAYLOAD",
        ));
    }

    let mut payload = vec![0u8; length];
    if length > 0 {
        r.read_exact(&mut payload).await?;
    }

    Ok(Some(Frame {
        typ,
        stream_id,
        payload,
    }))
}

/// Write one frame to `w`.
#[allow(dead_code)]
pub async fn write_frame<W: AsyncWrite + Unpin>(w: &mut W, frame: &Frame) -> io::Result<()> {
    let length = frame.payload.len() as u32;
    let mut hdr = [0u8; HEADER_LEN];
    hdr[0] = frame.typ;
    hdr[1..5].copy_from_slice(&frame.stream_id.to_be_bytes());
    hdr[5..9].copy_from_slice(&length.to_be_bytes());
    w.write_all(&hdr).await?;
    if !frame.payload.is_empty() {
        w.write_all(&frame.payload).await?;
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn rejects_oversized_frame_header() {
        // type=DATA, stream=1, length = MAX_FRAME_PAYLOAD + 1
        let mut bytes = vec![DATA];
        bytes.extend_from_slice(&1u32.to_be_bytes());
        bytes.extend_from_slice(&((MAX_FRAME_PAYLOAD + 1) as u32).to_be_bytes());
        let mut cursor = std::io::Cursor::new(bytes);
        assert!(read_frame(&mut cursor).await.is_err());
    }
}
